xxh3-dispatch.h

```diff
@@ -0,0 +1,277 @@
+
+        if (f != NULL) {
+            while (fgets(buf, 1024, f) != NULL) {
+                if (strstr(buf, "neon") != NULL || strstr(buf, "asimd") != NULL) {
+                    fclose(f);
+
+                    cpu_mode = supported_cpu_mode = XXH_CPU_MODE_NEON;
+                    XXH3_hashLong = &_XXH3_hashLong_NEON;
+                    return XXH3_hashLong;
+                }
+            }
+            fclose(f);
+        }
+    }
+#endif
+    /* Must be scalar. */
+    cpu_mode = supported_cpu_mode = XXH_CPU_MODE_SCALAR;
+    XXH3_hashLong = &_XXH3_hashLong_Scalar;
+    return XXH3_hashLong;
+}
+
+/* Sets up the dispatcher and then calls the actual hash function. */
+static void
+XXH3_dispatcher(U64* restrict acc, const void* restrict data, size_t len, const U32* restrict key)
+{
+    /* We haven't checked CPUID yet, so we check it now. On GCC, we try to get this to run
+     * at program startup to hide our very dirty secret from the benchmarks. */
+    return (XXH3_featureTest())(acc, data, len, key);
+}
+
+/* Sets the XXH3_hashLong variant. When XXH_MULTI_TARGET is not defined, this
+ * does nothing.
+ *
+ * Unless XXH_NO_VERIFY_MULTI_TARGET is defined, this will automatically fall back
+ * to the next best XXH3 mode, so, for example, even if you set it to AVX2, the code
+ * will not crash even if it is run on, for example, a Core 2 Duo which doesn't support
+ * AVX2. */
+XXH_PUBLIC_API XXH_cpu_mode_t XXH3_setCpuMode(XXH_cpu_mode_t mode)
+{
+/* Defining XXH_NO_VERIFY_MULTI_TARGET will allow you to set the CPU mode to
+ * an unsupported mode. */
+#ifndef XXH_NO_VERIFY_MULTI_TARGET
+    /* Call the featureTest if it hasn't been called already */
+    if (supported_cpu_mode == XXH_CPU_MODE_AUTO)
+        XXH3_featureTest();
+/* Even thougu XXH_CPU_MODE_NEON is greater than that, it will never
+ * be defined */
+#   define TRY_SET_MODE(mode_num, funcptr) \
+    do { \
+        if (mode == (mode_num) && supported_cpu_mode >= (mode_num)) { \
+            cpu_mode = (mode_num); \
+            XXH3_hashLong = &(funcptr); \
+            return (mode_num); \
+        } \
+   } while (0)
+#else
+#   define TRY_SET_MODE(mode_num, funcptr) \
+   do { \
+       if (mode == (mode_num)) { \
+            cpu_mode = (mode_num); \
+            XXH3_hashLong = &(funcptr); \
+            return (mode_num); \
+       }
+    } while (0)
+#endif
+
+#if XXH_TARGET_MODE == XXH_TARGET_MODE_X86
+    TRY_SET_MODE(XXH_CPU_MODE_AVX2, _XXH3_hashLong_AVX2);
+    TRY_SET_MODE(XXH_CPU_MODE_SSE2, _XXH3_hashLong_SSE2);
+#elif XXH_TARGET_MODE == XXH_TARGET_MODE_ARM
+    TRY_SET_MODE(XXH_CPU_MODE_NEON, _XXH3_hashLong_NEON);
+#endif
+    if (mode == XXH_CPU_MODE_SCALAR) {
+        cpu_mode = XXH_CPU_MODE_SCALAR;
+        XXH3_hashLong = &_XXH3_hashLong_Scalar;
+        return XXH_CPU_MODE_SCALAR;
+    }
+    cpu_mode = XXH_CPU_MODE_AUTO;
+    XXH3_hashLong = &XXH3_dispatcher;
+    return XXH_CPU_MODE_AUTO;
+#undef TRY_SET_MODE
+}
+
+/* Should we keep this? */
```

Comment:
That's a good question.
I definitely understand the function while in debugging mode.

Now, outside of this situation, this seems less essential, but why not.
The xxhash API has a category protected by macro guard `XXH_STATIC_LINKING_ONLY`.
I believe it could be present there.